refactor: move structure to a monorepo with multiple packages (#381)
cognitedata/cognitesdk-js: CONTRIBUTING.md

Contributing
============

Contributions are welcome, and this document details how changes can be made and submitted,

> and eventually included in a relase. We use monorepo tooling, and a git setup for CI/CD.a PATCH bump 

Reviewer: ```suggestion
and eventually included in a release. We use monorepo tooling, and a git setup for CI/CD PATCH bump 
```

Please note we have a [code of conduct](./CODE_OF_CONDUCT.md).

## Making changes
Make the changes to the package(s) you want to change, and commit them to a fork or branch.
Commits need to follow [proper commit messages](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines).
The commit messages are used to automatically bump the versions of changed packages, and write automatic changelogs.
We use semantic versioning, with versions `MAJOR.MINOR.PATCH`.

 - For fixes, start the commit with `fix:`. This will bump PATCH.
 - For features, start the commit with `feat:`. This will bump MINOR.
 - For breaking changes, add an exclamation mark, e.g. `feat!:` or `fix!:`. This will bump MAJOR.
 - For extra details in the changelog, you can specify a scope, e.g. `feat(assets):`.
 - For other changes, use `refactor:`, `test:`, `docs:`, `chore:` etc.

## Pull request
Make a pull request from your branch to the main branch. When merging the pull request,

> only use squashing if the resulting squash commit can accuratly describe the change as a single conventional commit.

Reviewer: ```suggestion
only use squashing if the resulting squash commit can accurately describe the change as a single conventional commit.
```

Once the change is pushed to the main branch, it is time for a release.

## Releases & Versioning
Releases are done from the main branch, so when a pull request is merged,
CI/CD will run tests, and if successful, do deploys.
Documentation is built and deployed, and code snippets
are exported to the service contract repo as a pull request.

Updating and uploading npm packages only happens if the HEAD commit of the main branch
contains `[release]` in its description. When CI/CD sees this, it will use lerna to update
package versions of changed packages based on commit messages, and add the 
changes to the changelogs. The changes are comitted to the main branch
with the new versions as git tags, and the new package versions are uploaded to npm.

> We restrict new npm releases to `[release]`-tagged commits becuase lerna is

Reviewer: ```suggestion
We restrict new npm releases to `[release]`-tagged commits because lerna is
```

> quite agressive in its versioning. Changes to any file not ignored by lerna will

Reviewer: ```suggestion
quite aggressive in its versioning. Changes to any file not ignored by lerna will
```

cause a PATCH bump. Markdown files and tests are ignored, but changing anything else,
like a comment in a source file, will trigger a new version,
irrespective of conventional commits.

This does *not* mean you should store unfinished work on the main branch.
Another package may be ready for release, and once a `[release]`
commit is pushed, all changed packages are updated.
Repository administrators should be in control of `[release]` commits.

To add a release commit to a clean working tree, use the command
```bash
git checkout v1
git pull
git commit --allow-empty -m "chore: trigger [release]"
git push
```

> Also keep in mind that the `[release]` commit has to be the HEAD of

Reviewer: ```suggestion
Also, keep in mind that the `[release]` commit has to be the HEAD of
```

main, and travis only runs on the HEAD. If HEAD has changed by the time

> the versiong happens, travis will fail.

Reviewer: ```suggestion
the versioning happens, travis will fail.
```

Once the change is pushed to the main branch, it is time for a release.

## Releases & Versioning
Releases are done from the main branch, so when a pull request is merged,
CI/CD will run tests, and if successful, do deploys.
Documentation is built and deployed, and code snippets
are exported to the service contract repo as a pull request.

Updating and uploading npm packages only happens if the HEAD commit of the main branch
contains `[release]` in its description. When CI/CD sees this, it will use lerna to update
package versions of changed packages based on commit messages, and add the 
changes to the changelogs. The changes are comitted to the main branch
with the new versions as git tags, and the new package versions are uploaded to npm.

We restrict new npm releases to `[release]`-tagged commits becuase lerna is
quite agressive in its versioning. Changes to any file not ignored by lerna will

> cause a PATCH bump. Markdown files and tests are ignored, but chaning anything else,

Reviewer 2: ```suggestion
cause a PATCH bump. Markdown files and tests are ignored, but changing anything else,
```

like a comment in a source file, will trigger a new version,
irrespective of conventional commits.

This does *not* mean you should store unfinished work on the main branch.
Another package may be ready for release, and once anyone pushes a `[release]`
commit, all changed packages are updated.

> Also keep in mind that the `[release]` commit has to be the HEAD of

Reviewer 2: maybe we just advice to trigger the push only for administrators?
so it won't be as a part of pr but rather the direct push

Reviewer 2: we have now a bit of a problem now with people merging to master without us knowing it

We restrict new npm releases to `[release]`-tagged commits becuase lerna is
quite agressive in its versioning. Changes to any file not ignored by lerna will
cause a PATCH bump. Markdown files and tests are ignored, but changing anything else,
like a comment in a source file, will trigger a new version,
irrespective of conventional commits.

This does *not* mean you should store unfinished work on the main branch.
Another package may be ready for release, and once a `[release]`
commit is pushed, all changed packages are updated.

> Repository administators should be in controll of `[release]` commits.

Reviewer 2: ```suggestion
Repository administrators should be in control of `[release]` commits.
```
